gem/optimise.py

```diff
@@ -413,3 +474,403 @@ def aggressive_unroll(expression):
+                if key == 'const' or key == 'other':
+                    continue
+                connection.extend([gem_int[factor] for factor in factors])
+            connections.append(tuple(connection))
+
+        # set up the ILP
+        import pulp as ilp
+        ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+        ilp_var = ilp.LpVariable.dicts('node', int_gem.keys(), 0, 1, ilp.LpBinary)
+
+        # Objective function
+        # Minimise number of factors to pull. If same number, favour factor with larger extent
+        big = 10000000  # some arbitrary big number
+        ilp_prob += ilp.lpSum(ilp_var[i] * (big - self.factor_extent(int_gem[i])) for i in int_gem)
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[i] for i in connection) >= 1
+
+        ilp_prob.solve()
+        if ilp_prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_factors = [factor for factor, number in iteritems(gem_int) if ilp_var[number].value() == 1]
+        other_factors = [factor for factor, number in iteritems(gem_int) if ilp_var[number].value() == 0]
+        # TODO: investigate effects of sorting these two lists of factors
+        optimal_factors = sorted(optimal_factors, key=lambda f: self.factor_extent(f), reverse=True)
+        other_factors = sorted(other_factors, key=lambda f: self.factor_extent(f), reverse=True)
+        # Sequence dictating order of factorisation
+        factors_seq = optimal_factors + other_factors
+        return factors_seq
+
+    def factorise_key(self, key):
+        """
+        Factorise common factors that have a particular :param key
+        """
+        # TODO: Add a forbidden list of factors
+        if len(self.rep) < 2:
+            return
+        counter = OrderedDict()
+        for summand in self.rep:
+            if key not in summand:
+                continue
+            for factor in summand[key]:
+                counter.setdefault(factor, 0)
+                counter[factor] += 1
+        if max(counter.values()) < 2:
+            return
+        saved_flops = OrderedDict((((count_flop(factor) + 1)) * (count - 1), factor)
+                                  for (factor, count) in iteritems(counter))
+        mcf = saved_flops[max(saved_flops.keys())]  # most common factor
+        self.opt_node = None
+        _summands = list()
+        factored_out = list()
+        for summand in self.rep:
+            # TODO: This pattern of checking keys is repeated several times
+            if key not in summand:
+                continue
+            if mcf in summand[key]:
+                factored_out.append(summand)  # mark for deleting later
+                _factors = OrderedDict([(_k, _v) for (_k, _v) in iteritems(summand) if _k != key])
+                _factors[key] = [factor for factor in summand[key] if factor != mcf]
+                _summands.append(_factors)
```

Comment:
This pattern, creating your sum-product tree IR appears over and over again.  What happens if you decide to use a different data structure?  You're screwed because there are so many places you need to change.

Think about how you want to represent this data, then maybe consider refactoring such that creating this representation is in a free function, or else you have a small helper object?  Our discussion on Tuesday was trying to reduce the use of this kind of deeply nested datastructure, I think.